portal/app/core/views/base.py

```diff
@@ -35,34 +35,34 @@ def index(request):
         template = 'partials/main_lab.html'
     elif request.user.powernetuser.type == PowernetUserType.FARM:
 
-        # querying power consumption data
-        # pen_power also used for bottom row control fan power
-        last_object = FarmData.objects.filter(farm_device_id = 17).latest('timestamp')
-        serialized_farm_data = FarmDataSerializer(last_object).data
-        pen_power = json.dumps(main_farm_parser(serialized_farm_data))
+        try:
+            # querying power consumption data
+            # pen_power also used for bottom row control fan power
+            last_object = FarmData.objects.filter(farm_device_id = 17).latest('timestamp')
+            serialized_farm_data = FarmDataSerializer(last_object).data
+            pen_power = json.dumps(main_farm_parser(serialized_farm_data))
+
+            # querying pen1 test data for the top row of fans
+            farm_device = FarmData.objects.filter(farm_device_id = 20).values('device_data').order_by('-timestamp')
+            serialized_farm_data = FarmDataSerializer(farm_device, many=True).data
+            pen1_fan = local_fan_info_parser(serialized_farm_data)
 
-        # querying pen1 test data for the top row of fans
-        farm_device = FarmData.objects.filter(farm_device_id = 20).values('device_data').order_by('-timestamp')
-        serialized_farm_data = FarmDataSerializer(farm_device, many=True).data
-        pen1_fan = local_fan_info_parser(serialized_farm_data)
-
-        # querying to get latest timestamp value
-        last_object = FarmData.objects.filter(farm_device_id__gte = 1,
-                                              farm_device_id__lte = 16).latest('timestamp')
-        serialized_last_object = FarmDataSerializer(last_object).data
-        last_timestamp = datetime.datetime.strptime(serialized_last_object['timestamp'],
-                                                    '%Y-%m-%dT%H:%M:%S.%fZ')
-        # last timestamp - 2 hours
-        time_2_hours_ago = last_timestamp - datetime.timedelta(hours=2)
-        # querying the last 2 hours of data for average humidity and average temperature
-        farm_device = FarmData.objects.filter(farm_device_id__gte = 1, farm_device_id__lte = 16,
-                                              timestamp__gte = time_2_hours_ago)
-        serialized_farm_data = FarmDataSerializer(farm_device, many=True).data
-        argv = ['temperature', 'rel_humidity', 'timestamp', serialized_farm_data]
-        temp_hum = energy_summary_parser(argv)
+            latest_ts = datetime.datetime.now()
+            # last timestamp - 2 hours
+            time_2_hours_ago = latest_ts - datetime.timedelta(hours=1)
+
+            # querying the last 2 hours of data for average humidity and average temperature
+            farm_device = FarmData.objects.filter(farm_device_id__gte = 1, farm_device_id__lte = 16, timestamp__gte = time_2_hours_ago)
```

Comment:
seems like you should be able to consolidate the `FarmData` queries into one instead of hitting the db 3 separate times

Reply:
Done

Reply:
Let me know what you think about the recent update

Reply:
lmk what you think